File: medium/Gas_Station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        diffs = [gas[i]-cost[i] for i in range(len(gas))]
        index = 0
        if sum(diffs) < 0:
            return -1
        queue = []
        nums = 0
        for i, diff in enumerate(diffs):
            nums+=diff
            if nums < 0:
                index = i+1
                queue.append(nums)
                nums = 0
        queue.append(nums)
        if sum(queue) >= 0:
            return index
        else:
            return -1

File: cpp/medium/Gas_Station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        vector<int> gap;
        for (int i = 0; i < gas.size(); ++i) {
            gap.push_back(gas[i] - cost[i]);
        }
        if (accumulate(gap.begin(), gap.end(), 0) < 0) return -1;
        int start = 0;
        int gas_amount = 0;
        for (int i = 0; i < gap.size(); ++i) {
            gas_amount = gas_amount + gap[i];
            if (gas_amount < 0) {
                start = i + 1;
                gas_amount = 0;
            }
        }
        return start;
    }
};
